File: ios/Application/ExtractArchiveUseCase.swift
import Foundation

struct ExtractionResult {
  let extractedEntries: Int
  let skippedEntries: Int
  let writtenBytes: UInt64
}

final class ExtractArchiveUseCase {
  private let engine: ArchiveEngine

  init(engine: ArchiveEngine) {
    self.engine = engine
  }

  func execute(
    session: ArchiveEngineSession,
    plan: [ExtractionPlanItem],
    destination: DirectoryOutput,
    limits: ExtractionLimits,
    password: String?,
    onProgress: @escaping (String, UInt64, UInt64) -> Void
  ) async throws -> ExtractionResult {
    var extracted = 0
    var skipped = 0
    var writtenBytes: UInt64 = 0

    for item in plan {
      try Task.checkCancellation()

      let path = item.outputPath.normalized
      let entry = item.entry

      if entry.kind == "directory" {
        try await destination.prepareDirectory(at: path)
        extracted += 1
        onProgress(path, 0, 0)
        continue
      }

      guard entry.kind == "file" else {
        skipped += 1
        continue
      }

      guard entry.uncompressedSize <= limits.maxEntryUncompressedBytes else {
        throw ArchiveDomainError.extractionLimitExceeded(
          limit: "maxEntryUncompressedBytes",
          value: "\(entry.uncompressedSize)"
        )
      }

      let data = try await engine.readEntry(
        session: session,
        path: entry.path,
        limit: limits.maxEntryUncompressedBytes,
        password: password
      )

      let fileURL = try await destination.createFile(at: path)
      try data.write(to: fileURL, options: .atomic)

      writtenBytes += UInt64(data.count)
      extracted += 1
      onProgress(path, UInt64(data.count), writtenBytes)
    }

    return ExtractionResult(
      extractedEntries: extracted,
      skippedEntries: skipped,
      writtenBytes: writtenBytes
    )
  }
}

File: android/src/main/java/com/mcodex/nitroarchive/application/ExtractArchiveUseCase.kt
package com.mcodex.nitroarchive.application

import com.mcodex.nitroarchive.application.ports.*
import com.mcodex.nitroarchive.domain.ArchiveDomainError
import com.mcodex.nitroarchive.domain.ExtractionLimits

data class ExtractionSummary(
    val extractedEntries: Int,
    val skippedEntries: Int,
    val writtenBytes: Long
)

class ExtractArchiveUseCase(private val engine: ArchiveEngine) {

    suspend fun execute(
        session: ArchiveEngineSession,
        entries: List<EntryDescriptor>,
        output: DirectoryOutput,
        limits: ExtractionLimits = ExtractionLimits.DEFAULT,
        password: String? = null,
        onEntryProgress: (path: String, index: Int, total: Int) -> Unit = { _, _, _ -> }
    ): ExtractionSummary {
        var extracted = 0
        var skipped = 0
        var written = 0L

        for ((i, descriptor) in entries.withIndex()) {
            onEntryProgress(descriptor.path, i, entries.size)

            val parentDir = if (descriptor.path.contains("/")) {
                descriptor.path.substringBeforeLast("/")
            } else ""

            when (descriptor.kind) {
                "directory" -> {
                    output.prepareDirectory(at = descriptor.path)
                    extracted++
                }
                "file" -> {
                    if (descriptor.uncompressedSize > limits.maxEntryUncompressedBytes) {
                        throw ArchiveDomainError.ExtractionLimitExceeded(
                            limit = "maxEntryUncompressedBytes",
                            value = "${descriptor.uncompressedSize}"
                        )
                    }

                    if (parentDir.isNotEmpty()) {
                        output.prepareDirectory(at = parentDir)
                    }
                    val file = output.createFile(at = descriptor.path)

                    val bytesWritten = engine.extractEntry(
                        session = session,
                        path = descriptor.path,
                        destination = file,
                        limit = limits.maxEntryUncompressedBytes,
                        password = password
                    )

                    extracted++
                    written += bytesWritten
                }
                else -> skipped++
            }
        }

        return ExtractionSummary(
            extractedEntries = extracted,
            skippedEntries = skipped,
            writtenBytes = written
        )
    }
}
